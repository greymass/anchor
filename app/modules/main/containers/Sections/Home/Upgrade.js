// @flow
import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { withTranslation } from 'react-i18next';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { Button, Dimmer, Header, Loader, Modal, Table } from 'semantic-ui-react';
import compose from 'lodash/fp/compose';

import { changeModule } from '../../../actions/navigation';
import { setSetting } from '../../../../../shared/actions/settings';
import { swapBlockchain } from '../../../../../shared/actions/blockchains';
import { upgradeV1Wallets } from '../../../../../shared/actions/wallets';

import GlobalFragmentChainLogo from '../../../../../shared/components/Global/Fragment/ChainLogo';
import GlobalButtonElevate from '../../../../../shared/containers/Global/Button/Elevate';
import GlobalFragmentAuthorization from '../../../../../shared/components/Global/Fragment/Authorization';

class HomeUpgradeContainer extends Component<Props> {
  onClose = () => {
    const {
      history,
    } = this.props;
    history.push('/');
  }
  onClick = (password) => {
    const {
      actions,
      upgradable
    } = this.props;
    actions.upgradeV1Wallets(upgradable, password);
  }
  render() {
    const {
      completed,
      system,
      t,
      total,
      upgradable,
    } = this.props;
    const loading = (system.V1UPGRADE === 'PENDING' || system.V1UPGRADE === 'PROGRESS');
    return (
      <React.Fragment>
        {completed}/{total}
        <Dimmer active={loading}>
          <Loader
            size="huge"
            style={{
              color: 'white'
            }}
          >
            <Header
              style={{
                color: 'white'
              }}
            >
              <Header.Subheader
                style={{
                  color: 'white'
                }}
              >
                {t('main_sections_home_upgrade_subheader_one')}
              </Header.Subheader>
              {t('main_sections_home_upgrade_header_one')}
            </Header>
          </Loader>
        </Dimmer>
        <Modal
          centered={false}
          open={!loading}
        >
          <Modal.Header>
            <Header
              content={t('main_sections_home_upgrade_header_two')}
              subheader={t('main_sections_home_upgrade_subheader_two')}
              style={{ marginTop: 0 }}
            />
          </Modal.Header>
          <Modal.Content>
            <p>
              <GlobalButtonElevate
                onSuccess={(password) => this.onClick(password)}
                trigger={(
                  <Button
                    content={t('main_sections_home_upgrade_button_one')}
                    primary
                  />
                )}
                wallet={upgradable[0]}
              />
            </p>
            <Table celled>
              {upgradable
                .sort((a, b) => a.account > b.account ? 1 : -1)
                .map((wallet) => (
                  <Table.Row>
                    <Table.Cell collapsing>
                      <GlobalFragmentChainLogo
                        avatar
                        chainId={wallet.chainId}
                      />
                    </Table.Cell>
                    <Table.Cell collapsing>
                      <GlobalFragmentAuthorization
                        account={wallet.account}
                        authorization={wallet.authorization}
                        pubkey={wallet.pubkey}
                      />
                    </Table.Cell>
                    <Table.Cell>
                      {t('main_sections_home_upgrade_table_cell_one', { walletMode: wallet.mode  })}
                    </Table.Cell>
                </Table.Row>
              ))}
            </Table>
          </Modal.Content>
        </Modal>
      </React.Fragment>
    );
  }
}

function mapStateToProps(state) {
  return {
    completed: (state.system.V1UPGRADE_PROGRESS) ? state.system.V1UPGRADE_PROGRESS.completed : 0,
    settings: state.settings,
    system: state.system,
    total: (state.system.V1UPGRADE_PROGRESS) ? state.system.V1UPGRADE_PROGRESS.total : 0,
    upgradable: state.wallets.filter(w => w.data),
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      changeModule,
      setSetting,
      swapBlockchain,
      upgradeV1Wallets,
    }, dispatch)
  };
}

export default compose(
  withRouter,
  withTranslation('main'),
  connect(mapStateToProps, mapDispatchToProps)
)(HomeUpgradeContainer);
