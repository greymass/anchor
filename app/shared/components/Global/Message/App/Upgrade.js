// @flow
import React, { Component } from 'react';
import { translate } from 'react-i18next';

import { Button, Form, Header, Icon, Segment } from 'semantic-ui-react';

const { ipcRenderer } = require('electron');

export class GlobalMessageAppUpgrade extends Component<Props> {
  checkForUpdates = () => {
    ipcRenderer.send('checkForUpdates');
  };
  skipUpdate = () => {
    const {
      actions,
      constants
    } = this.props;
    actions.setSetting('upgradeSkip', constants.version);
  };
  render() {
    const {
      constants,
      t,
    } = this.props;
    return (
      <React.Fragment>
        <Segment basic size="large" textAlign="center">
          <Header color="blue" icon size="large">
            <Icon name="info circle" />
            <Header.Content>
              <React.Fragment>
                {t('global_app_requires_upgrade_header', { new: constants.version })}
                <Header.Subheader>
                  <p>
                    {t('global_app_requires_upgrade_subheader')}
                  </p>
                </Header.Subheader>
              </React.Fragment>
            </Header.Content>
          </Header>
          <Form>
            <Form.TextArea
              label={t('global_app_requires_upgrade_details')}
              rows={6}
              value={constants.versioninfo[constants.version]}
            />
          </Form>
        </Segment>
        <Segment basic size="large" textAlign="center">
          <p>
            <Button
              content="Upgrade"
              color="blue"
              icon="up arrow"
              onClick={this.checkForUpdates}
              size="large"
            />
          </p>
          <p>
            <Button
              content="Skip this Version"
              color="grey"
              onClick={this.skipUpdate}
            />
          </p>
        </Segment>
      </React.Fragment>
    );
  }
}

export default translate('global')(GlobalMessageAppUpgrade);
