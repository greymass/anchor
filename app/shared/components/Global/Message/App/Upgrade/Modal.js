// @flow
import React, { Component } from 'react';
import { translate } from 'react-i18next';

import { Modal } from 'semantic-ui-react';

import GlobalMessageAppUpgrade from '../Upgrade';

export class GlobalMessageAppUpgradeModal extends Component<Props> {
  render() {
    const {
      actions,
      constants,
      name,
      version
    } = this.props;
    return (
      <Modal
        content={(
          <GlobalMessageAppUpgrade
            actions={actions}
            constants={constants}
            name={name}
            version={version}
          />
        )}
        open
        size="small"
      />
    );
  }
}

export default translate('global')(GlobalMessageAppUpgradeModal);
