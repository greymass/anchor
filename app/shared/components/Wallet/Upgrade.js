// @flow
import React, { Component } from 'react';
import { translate } from 'react-i18next';

import { Header, Icon, Segment } from 'semantic-ui-react';

import GlobalButtonWalletUpgrade from '../../containers/Global/Button/Wallet/Upgrade';

class WalletUpgrade extends Component<Props> {
  render() {
    const {
      settings,
      t,
      wallet
    } = this.props;
    const {
      skipImport,
      walletTemp
    } = settings;
    const needsUpgrade = (!wallet.pubkey || !wallet.authorization);
    if (!needsUpgrade || walletTemp || skipImport) return false;
    return (
      <Segment attached stacked>
        <Header>
          <GlobalButtonWalletUpgrade
            floated="left"
            style={{ marginRight: '1em' }}
            useWallet
            wallet={wallet}
          />
          <Icon name="warning sign" />
          <Header.Content>
            {t('wallet_requires_upgrade_header', { account: wallet.account })}
            <Header.Subheader>
              {t('wallet_requires_upgrade_subheader')}
            </Header.Subheader>
          </Header.Content>
        </Header>
      </Segment>
    );
  }
}

export default translate('wallet')(WalletUpgrade);
