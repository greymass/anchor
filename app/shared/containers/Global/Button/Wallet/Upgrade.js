// @flow
import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { translate } from 'react-i18next';
import compose from 'lodash/fp/compose';
import { Button, Dropdown, Modal, Popup, Segment } from 'semantic-ui-react';

import GlobalButtonElevate from '../Elevate';
import * as WalletsActions from '../../../../actions/wallets';

class GlobalButtonWalletUpgrade extends Component<Props> {
  state = {
    current: {
      account: false,
      authorization: false,
    },
    open: false
  }
  selectWatchWalletAuthorization = (account, authorization) => {
    this.setState({
      current: {
        account,
        authorization
      },
      open: true,
    });
  }
  setAuthorization = (e, { value }) => {
    const { current } = this.state;
    this.setState({
      current: {
        account: current.account,
        authorization: value
      }
    });
  }
  upgradeWallet = (account, authorization, password = false) => {
    const { actions, useWallet } = this.props;
    actions.upgradeWallet(account, authorization, password, useWallet);
  }
  upgradeWatchWallet = (account, authorization) => {
    const { actions, useWallet } = this.props;
    actions.upgradeWatchWallet(account, authorization, useWallet);
  }
  render() {
    const {
      floated,
      style,
      settings,
      t,
      validate,
      wallet
    } = this.props;
    const {
      account,
      authorization,
      mode,
      pubkey
    } = wallet;
    const {
      current,
      open
    } = this.state;
    const requiresUpgrade = (!authorization || !pubkey);
    // Cold wallets don't use this feature
    if (mode === 'cold') return false;
    let modal;
    if (open) {
      const options = ['active', 'owner'].map((authority) => (
        {
          key: authority,
          text: authority,
          value: authority
        }
      ));
      modal = (
        <Modal
          open
          size="small"
        >
          <Modal.Header>
            {t('wallet:wallet_upgrade_watch_authorization_header')}
          </Modal.Header>
          <Modal.Content>
            <p>{t('wallet:wallet_upgrade_watch_authorization_description')}</p>
            <Dropdown
              fluid
              onChange={this.setAuthorization}
              options={options}
              selection
            />
            <Segment basic clearing>
              <Button
                color="purple"
                content={t('wallet:wallet_upgrade_button')}
                disabled={!current.authorization}
                floated="right"
                icon="circle arrow up"
                onClick={() => this.upgradeWatchWallet(current.account, current.authorization)}
                style={style}
              />
            </Segment>
          </Modal.Content>
        </Modal>
      );
    }
    return (
      <React.Fragment>
        {modal}
        {(requiresUpgrade && mode === 'watch')
          ? (
            <Popup
              content={t('wallet:wallet_upgrade_popup')}
              inverted
              trigger={(
                <Button
                  color="purple"
                  content={t('wallet:wallet_upgrade_button')}
                  floated={floated}
                  icon="circle arrow up"
                  onClick={() => this.selectWatchWalletAuthorization(account, authorization)}
                  style={style}
                />
              )}
            />
          )
          : false
        }
        {(requiresUpgrade && mode !== 'watch')
          ? (
            <Popup
              content={t('wallet:wallet_upgrade_popup')}
              inverted
              trigger={(
                <GlobalButtonElevate
                  onSuccess={(password) =>
                    this.upgradeWallet(account, authorization, password)
                  }
                  settings={settings}
                  trigger={(
                    <Button
                      color="purple"
                      content={t('wallet:wallet_upgrade_button')}
                      floated={floated}
                      icon="circle arrow up"
                      style={style}
                    />
                  )}
                  validate={validate}
                  wallet={wallet}
                />
              )}
            />
          )
          : false
        }
      </React.Fragment>
    );
  }
}

function mapStateToProps(state) {
  return {
    settings: state.settings,
    validate: state.validate
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      ...WalletsActions,
    }, dispatch)
  };
}

export default compose(
  translate('global'),
  connect(mapStateToProps, mapDispatchToProps)
)(GlobalButtonWalletUpgrade);
