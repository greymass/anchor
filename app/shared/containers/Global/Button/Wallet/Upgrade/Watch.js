// @flow
import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { translate } from 'react-i18next';
import compose from 'lodash/fp/compose';
import { Button, Dropdown, Modal, Popup } from 'semantic-ui-react';

import GlobalButtonElevate from '../Elevate';
import * as WalletsActions from '../../../../actions/wallets';

class GlobalButtonWalletUpgradeWatch extends Component<Props> {
  state = {
    current: {
      account: 'teamgreymass',
      authorization: false,
    },
    open: true
  }
  selectWatchWalletAuthorization = (account, authorization) => {
    this.setState({
      current: {
        account,
        authorization
      },
      open: true,
    });
  }
  setAuthorization = (e, { value }) => {
    const { current } = this.state;
    this.setState({
      current: {
        account: current.account,
        authorization: value
      }
    });
  }
  upgradeWallet = (account, authorization, password = false) => {
    const { actions, useWallet } = this.props;
    actions.upgradeWallet(account, authorization, password, useWallet);
  }
  upgradeWatchWallet = (account, authorization) => {
    const { actions, useWallet } = this.props;
    actions.upgradeWatchWallet(account, authorization, useWallet);
  }
  render() {
    const {
      floated,
      style,
      settings,
      t,
      useWallet,
      validate,
      wallet
    } = this.props;
    const {
      account,
      authorization,
      mode,
      pubkey
    } = wallet;
    const {
      current,
      open
    } = this.state;
    const requiresUpgrade = (!authorization || !pubkey);
    let modal;
    if (open) {
      const options = ['active', 'owner'].map((authority) => (
        {
          key: authority,
          text: authority,
          value: authority
        }
      ));
      modal = (
        <GlobalButtonWalletUpgradeWatch
          account={current.account}
          authorization={current.authorization}
          onSubmit={() => this.upgradeWatchWallet(current.account, current.authorization)}
          open={open}
          setAuthorization={this.setAuthorization}
        />

        <Modal.Content>
          <p>Account: {current.account}</p>
          <p>authorization: {current.authorization}</p>
          <Dropdown
            onChange={this.setAuthorization}
            options={options}
            selection
          />
          <Button
            color="purple"
            content={t('wallet:wallet_upgrade_button')}
            floated="right"
            icon="circle arrow up"
            onClick={() => this.upgradeWatchWallet(current.account, current.authorization)}
            style={style}
          />
        </Modal.Content>
      </Modal>
      );
    }
    return (
      <React.Fragment>
        {modal}
        {(requiresUpgrade && mode === 'watch')
          ? (
            <Popup
              content={t('wallet:wallet_upgrade_popup')}
              inverted
              trigger={(
                <Button
                  color="purple"
                  content={t('wallet:wallet_upgrade_button')}
                  floated={floated}
                  icon="circle arrow up"
                  onClick={() => this.selectWatchWalletAuthorization(account, authorization)}
                  style={style}
                />
              )}
            />
          )
          : false
        }
        {(requiresUpgrade && mode !== 'watch')
          ? (
            <Popup
              content={t('wallet:wallet_upgrade_popup')}
              inverted
              trigger={(
                <GlobalButtonElevate
                  onSuccess={(password) =>
                    this.upgradeWallet(account, authorization, password)
                  }
                  settings={settings}
                  trigger={(
                    <Button
                      color="purple"
                      content={t('wallet:wallet_upgrade_button')}
                      floated={floated}
                      icon="circle arrow up"
                      style={style}
                    />
                  )}
                  validate={validate}
                  wallet={wallet}
                />
              )}
            />
          )
          : false
        }
      </React.Fragment>
    );
  }
}

function mapStateToProps(state) {
  return {
    settings: state.settings,
    validate: state.validate
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      ...WalletsActions,
    }, dispatch)
  };
}

export default compose(
  translate('global'),
  connect(mapStateToProps, mapDispatchToProps)
)(GlobalButtonWalletUpgradeWatch);
