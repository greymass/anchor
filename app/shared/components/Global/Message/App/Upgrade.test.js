import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { GlobalMessageAppUpgrade } from './Upgrade';

const mockProps = {
  actions: {},
  constants: { version: 1, versioninfo: { 1: 'constant' } },
  t: jest.fn,
};

describe('GlobalMessageAppUpgrade', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<GlobalMessageAppUpgrade {...mockProps} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
