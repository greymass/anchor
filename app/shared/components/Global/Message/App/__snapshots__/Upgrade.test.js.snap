// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalMessageAppUpgrade renders correctly 1`] = `
<Fragment>
  <Segment
    basic={true}
    size="large"
    textAlign="center"
  >
    <Header
      color="blue"
      icon={true}
      size="large"
    >
      <Icon
        as="i"
        name="info circle"
      />
      <HeaderContent>
        <Component />
        <HeaderSubheader>
          <p>
            <Component />
          </p>
        </HeaderSubheader>
      </HeaderContent>
    </Header>
    <Form
      as="form"
    >
      <FormTextArea
        as={[Function]}
        control={[Function]}
        label={[MockFunction]}
        rows={6}
        value="constant"
      />
    </Form>
  </Segment>
  <Segment
    basic={true}
    size="large"
    textAlign="center"
  >
    <p>
      <Button
        as="button"
        color="blue"
        content="Upgrade"
        icon="up arrow"
        onClick={[Function]}
        size="large"
      />
    </p>
    <p>
      <Button
        as="button"
        color="grey"
        content="Skip this Version"
        onClick={[Function]}
      />
    </p>
  </Segment>
</Fragment>
`;
